File: services/auth-service/src/utils/password.ts
import bcrypt from 'bcryptjs';
import { config } from '../config';
import { logger } from './logger';

export class PasswordService {
  /**
   * Hash a password using bcrypt
   */
  static async hashPassword(password: string): Promise<string> {
    try {
      const salt = await bcrypt.genSalt(config.bcrypt.rounds);
      const hashedPassword = await bcrypt.hash(password, salt);
      logger.debug('Password hashed successfully');
      return hashedPassword;
    } catch (error) {
      logger.error('Password hashing failed:', error);
      throw new Error('Failed to hash password');
    }
  }

  /**
   * Verify a password against its hash
   */
  static async verifyPassword(password: string, hashedPassword: string): Promise<boolean> {
    try {
      const isValid = await bcrypt.compare(password, hashedPassword);
      logger.debug('Password verification completed');
      return isValid;
    } catch (error) {
      logger.error('Password verification failed:', error);
      throw new Error('Failed to verify password');
    }
  }

  /**
   * Validate password strength
   */
  static validatePasswordStrength(password: string): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (password.length < 8) {
      errors.push('Password must be at least 8 characters long');
    }

    if (!/[A-Z]/.test(password)) {
      errors.push('Password must contain at least one uppercase letter');
    }

    if (!/[a-z]/.test(password)) {
      errors.push('Password must contain at least one lowercase letter');
    }

    if (!/\d/.test(password)) {
      errors.push('Password must contain at least one number');
    }

    if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password)) {
      errors.push('Password must contain at least one special character');
    }

    return {
      isValid: errors.length === 0,
      errors,
    };
  }
}

File: backend/app/utils/password.py
"""
Password hashing and verification utilities
Uses bcrypt for secure password hashing with configurable salt rounds
"""
from passlib.context import CryptContext
from app.core.config import settings
from app.utils.error_sanitizer import get_generic_password_error
import logging

logger = logging.getLogger(__name__)


# Password context for bcrypt hashing with configured rounds
# Requirement 2.1: Use bcrypt with minimum 12 salt rounds
pwd_context = CryptContext(
    schemes=["bcrypt"],
    deprecated="auto",
    bcrypt__rounds=settings.BCRYPT_ROUNDS
)


def validate_password_config() -> None:
    """
    Validate password hashing configuration on service startup
    
    Raises:
        ValueError: If bcrypt rounds is less than 12 (security requirement)
    """
    if settings.BCRYPT_ROUNDS < 12:
        raise ValueError(
            f"BCRYPT_ROUNDS must be at least 12 for security (Requirement 2.1). "
            f"Current value: {settings.BCRYPT_ROUNDS}"
        )


def hash_password(password: str) -> str:
    """
    Hash a plain text password using bcrypt with configured salt rounds
    
    Security features:
    - Uses bcrypt algorithm (resistant to GPU attacks)
    - Configurable salt rounds (minimum 12, default 12)
    - Automatic salt generation
    - Secure error handling (Requirement 2.5)
    
    Args:
        password: Plain text password
        
    Returns:
        Hashed password string
        
    Raises:
        ValueError: If password hashing fails (with sanitized error message)
        
    Note:
        Requirement 2.1: Uses bcrypt with minimum 12 salt rounds
        Requirement 2.5: Returns generic error messages without exposing implementation details
    """
    try:
        return pwd_context.hash(password)
    except Exception as e:
        # Log the actual error for debugging (server-side only)
        logger.error(f"Password hashing failed: {type(e).__name__}", exc_info=False)
        
        # Return sanitized error message to user
        # Never expose bcrypt details, salt info, or internal errors
        raise ValueError(get_generic_password_error()) from None


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """
    Verify a plain text password against a hashed password
    
    Security features:
    - Uses constant-time comparison (prevents timing attacks)
    - Handles invalid hash gracefully
    - Never exposes error details to prevent information disclosure
    
    Args:
        plain_password: Plain text password to verify
        hashed_password: Hashed password to compare against
        
    Returns:
        True if password matches, False otherwise
        
    Note:
        Requirement 2.2: Uses constant-time comparison to prevent timing attacks
        Requirement 2.5: Handles errors gracefully without exposing implementation details
    """
    try:
        return pwd_context.verify(plain_password, hashed_password)
    except Exception as e:
        # Log the error type for debugging (server-side only)
        logger.warning(f"Password verification failed: {type(e).__name__}")
        
        # Return False instead of raising - don't expose error details
        # This prevents information disclosure about hash format, bcrypt version, etc.
        return False


def validate_password_strength(password: str) -> tuple[bool, str]:
    """
    Validate password strength requirements
    - Minimum 8 characters
    - At least one uppercase letter
    - At least one lowercase letter
    - At least one digit
    - At least one special character
    
    Args:
        password: Password to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if len(password) < 8:
        return False, "Password must be at least 8 characters long"
    
    if not any(char.isupper() for char in password):
        return False, "Password must contain at least one uppercase letter"
    
    if not any(char.islower() for char in password):
        return False, "Password must contain at least one lowercase letter"
    
    if not any(char.isdigit() for char in password):
        return False, "Password must contain at least one digit"
    
    special_chars = "!@#$%^&*()_+-=[]{}|;:,.<>?"
    if not any(char in special_chars for char in password):
        return False, "Password must contain at least one special character"
    
    return True, ""
